classes/calendar_signup.php

```diff
@@ -0,0 +1,112 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Calendar view to signup to timeslots.
+ *
+ * @package   mod_observation
+ * @copyright  2021 Endurer Solutions Team
+ * @author Matthew Hilton <mj.hilton@outlook.com>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace mod_observation;
+
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Calendar view to signup to timeslots.
+ *
+ * @package   mod_observation
+ * @copyright  2021 Endurer Solutions Team
+ * @author Matthew Hilton <mj.hilton@outlook.com>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class calendar_signup {
+
+    /**
+     * Generates a calendar signup HTML block.
+     * @param int $observationid ID of the observation instance
+     * @param string $title calendar title
+     * @param int $month month to show (1-12)
+     * @param int $year year to show
+     * @param string $baseurl URL used in button callbacks
+     */
+    public static function calendar_signup_view(int $observationid, string $title, int $month, int $year, string $baseurl) {
+        global $OUTPUT;
+
+        // Get the relevant timeslots.
+        $calendarslots = \mod_observation\timeslot_manager::get_timeslots_for_calendar($observationid, $month, $year);
+
+        $monthdays = array_fill(0, 31, []);
+
+        $monthdays = array_map(function($month, $i) {
+            return (object) [
+                "daynum" => $i + 1,
+                "events" => []
+            ];
+        }, $monthdays, array_keys($monthdays));
+
+        // Fill the array.
+        foreach ($calendarslots as $slot) {
+            $timeslotday = (int) date('j', $slot->start_time) - 1;
+
+            // Add extra data to use in the template.
+            $slot->start_time_formatted = userdate($slot->start_time);
+            $slot->join_button = $OUTPUT->single_button(new \moodle_url($baseurl, ['id' => $slot->obs_id,
+                'slotid' => $slot->id, 'action' => 'join', 'sesskey' => sesskey()]), get_string('join', 'observation'));
+
+            array_push($monthdays[$timeslotday]->events, $slot);
+        }
+
+        $weeks = array_fill(0, 5, []);
+
+        // Split into weeks.
```

Comment:
I think this is close, but I think first you need to find the day of the week, and pad it with empty cells for the first x days. Eg friday would have 5 days of padding (sunday first), which would then put each weekday in the same column all the way down (I think)

Also I think put a little M for monday next to the date inside the cell view